### PR TITLE
test: Add property tests for Dec type (#242)

### LINKED ISSUE #242: Update math package and add more thorough tests
We are using APD package, but we don't test it and it also turns out that it's not maintained.
We have to ensure that:
* it works as expected
* we use it correctly

### Moreover:
There are issues when we copy APD Decimals. We should prohibit copying APD Decimals
References - look at the related work and maybe port it:
* https://github.com/regen-network/mainnet/pull/15 
* https://github.com/regen-network/regen-ledger/pull/307

### PR DESCRIPTION
**Pull request by ruhatch**
## Description

Add `rapid` property tests for the `Dec` type

Closes #242
Closes #243

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: types/math/dec_test.go [hunk 1]
```diff
@@ -89,28 +132,402 @@ func TestDec(t *testing.T) {
...
+	require.NoError(t, err)
+
+	d, err := c.Sub(b)
+	require.NoError(t, err)
+
+	require.True(t, a.IsEqual(d))
+}
+
+// Property: a * (b + c) == (a * b) + (a * c)
+func testAddMulDistributes(t *rapid.T) {
```
Comment (ruhatch) [inline types/math/dec_test.go:341]: So right now we can generate arbitrarily large errors with this one. I'm not sure how we can constrain the inputs or operations in some way that we can at least provide some guarantees.

**Review by technicallyty — COMMENTED** [review]
not sure how in depth we want to go with these here but i think there a few more simple properties that we could add here:

- `a * 0 = 0`
- `a/1 = a`
- `a/a = 1`

File: types/math/dec_test.go [hunk 2]
```diff
@@ -89,28 +137,437 @@ func TestDec(t *testing.T) {
...
+	float float64
+	dec   Dec
+}
+
+// Generate a Dec value along with the float used to create it
+var genFloatAndDec *rapid.Generator = rapid.Custom(func(t *rapid.T) floatAndDec {
+	f := rapid.Float64().Draw(t, "f").(float64)
+	dec, err := NewDecFromString(fmt.Sprintf("%g", f))
+	if err != nil {
+		t.Fatalf("Failed while generating Dec from float: %g", f)
```
Comment (amaurym) [inline types/math/dec_test.go:163]: question: is there a reason why we don't do `require.NoError(t, err, "...")` here?

File: types/math/dec_test.go [hunk 3]
```diff
@@ -89,28 +137,437 @@ func TestDec(t *testing.T) {
...
+	c, err := a.Add(b)
+	require.NoError(t, err)
+
+	d, err := c.Sub(b)
+	require.NoError(t, err)
+
+	require.True(t, a.IsEqual(d))
+}
+
+// TODO: Revive this property test with appropriate errors once we need Mul
```
Comment (amaurym) [inline types/math/dec_test.go:345]: I generally don't like to commit commented code, would rather create an issue and paste this code there.

But also won't block this PR either.

File: types/math/dec_test.go [hunk 4]
```diff
@@ -89,28 +137,437 @@ func TestDec(t *testing.T) {
...
+
+	b, err := a.Quo(a)
+	require.NoError(t, err)
+	require.True(t, one.IsEqual(b))
+}
+
+// Property: a/1 = a
+func testQuoByOne(t *rapid.T) {
+	decNotZero := func(d Dec) bool { return !d.IsZero() }
+	a := genDec.Filter(decNotZero).Draw(t, "a").(Dec)
```
Comment (amaurym) [inline types/math/dec_test.go:450]: we don't need the filter for this test

**Review by amaurym — APPROVED** [review]
nice!

File: types/math/dec_test.go (hunk 2 above)
Comment (technicallyty) [reply types/math/dec_test.go:163]: i think this was for debugging purposes - I think we can safely remove it now

File: types/math/dec_test.go (hunk 4 above)
Comment (technicallyty) [reply types/math/dec_test.go:450]: ah right, got mixed up between 0/x and x/0

File: types/math/dec_test.go (hunk 3 above)
Comment (technicallyty) [reply types/math/dec_test.go]: https://github.com/regen-network/regen-ledger/issues/471

**Comment by clevinson** [comment]
🎉

**Pull request merged**